Fix public download activity

come-nc opened the pull request:
* Resolves: #51097 

## Summary

Fix pubilc download activity in 31+, and improve upon it to make sure it’s consistent.

## TODO

- [x] Viewing a video triggers several activity events. They are merged in activity feed, but generate a lot of notifications.
- [ ] ~Viewing a photo through preview does not trigger activity, is that an issue?~
- [x] Downloading the whole share triggers activity for the folder and then for each file.
- [x] Downloading several files in a zip triggers activity for each file. I think that’s good?
- [x] The manual activity publishing in ShareController should be removed to avoid duplicates. But I think that’s the one publishing the activity for folder download, so we should make sure we do not lose it.

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

come-nc committed:
fix(sharing): Publish activity for download by public link

Even if dav endpoint is now used.

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

susnux commented on apps/files_sharing/lib/Listener/BeforeNodeReadListener.php:
```diff
@@ -0,0 +1,116 @@
+use OCP\Files\Events\Node\BeforeNodeReadEvent;
+use OCP\Files\File;
+use OCP\Files\IRootFolder;
+use OCP\Files\Storage\ISharedStorage;
```
```suggestion
use OCP\Files\NotFoundException;
use OCP\Files\Storage\ISharedStorage;
```

come-nc committed:
fix: Fix download activity for folders

Remove duplicate activity publishing from share controller download,
 listen to BeforeZipCreatedEvent to publish activity for folders, and
 cache folders activity to avoid sending activity for each file in the
 folder.

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
fix: Avoid triggering several activities for Range request on the same file

Avoids flooding activities when someone browse a video in the web player.

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented:
> * [ ]  Viewing a photo through preview does not trigger activity, is that an issue?

This one is not a regression so let’s postpone for later.
A solution would be to listen to `BeforePreviewFetchedEvent`, but we’d have to decide at which size of preview we consider that downloading, because opening the file list already loads previews of small sizes.

come-nc commented:
/backport to stable31

come-nc committed:
fix(files_sharing): Set higher priority for listeners that may abort the request

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
fix(files_sharing): Use session id as part of cache key to avoid concurrency issues

If several people are watching and seeking the same video file we do not
 want the cache key to be the same or it would flood activity again.

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented:
I would like to add tests for this, but activity application is needed.
I can add the tests in activity application cypress tests, but that mean a PR breaking the feature on server will have green CI and we will only figure it out later. Still better than no tests I suppose?
Other ideas?

susnux commented:
> I would like to add tests for this, but activity application is needed.

Is there no OCS API or something you can test with behat?

Otherwise I see two ways:
1. Do it like we do for some Talk tests and install the Activity for some Behat tests (IMHO this makes sense for all things that have integrations in OCP)
2. Only add them to Activity -> would not be directly spotted if something breaks the app.

daniel-richter commented:
Just a heads-up: I had the same issue (downloads of individual files from a shared folder weren't showing up in the activities, #51097). I can confirm that the suggested changes solve the problem for me.

daniel-richter commented:
> * [x]  Viewing a video triggers several activity events. They are merged in activity feed, but generate a lot of notifications.
> * [ ]  ~Viewing a photo through preview does not trigger activity, is that an issue?~

Well, that's how it used to be. If you watched a video, it would show up in the "downloaded via public link" log. For images, you had to click "download" for it to show up in the log. Just viewing the image didn't do it.

susnux commented:
Did not yet had time to test this, but old behavior was also inconsistent as there were some endpoints that never caused a download notification.
Also we should define what this notification means.

Because either it only means real downloads, then we only need to track the DAV endpoints (`GET` on file or zipfolder plugin).
Or it also means viewing the file, then we also need to track preview endpoints (which is used by the viewer for images).

If I get the code right then this PR implements both. Which is fine but might confuse users as they will see "file downloaded" notifications even if they set "download permission" to `false` because previews are shown.

daniel-richter commented:
I see two options here:
1. Track only real _downloads_.
2. If we want to track _views_, too, we should introduce a separate subject, "public_shared_file_viewed" (in contrast to "public_shared_file_downloaded").

come-nc commented:
> > I would like to add tests for this, but activity application is needed.
> 
> Is there no OCS API or something you can test with behat?

It’s DAV, not OCS, but that can be tested with behat. But still needs the activity app.

> Otherwise I see two ways:
> 
> 1. Do it like we do for some Talk tests and install the Activity for some Behat tests (IMHO this makes sense for all things that have integrations in OCP)

I will try that.

> If I get the code right then this PR implements both. Which is fine but might confuse users as they will see "file downloaded" notifications even if they set "download permission" to `false` because previews are shown.

No, this PR only logs downloads through DAV, kind of the same as old behavior.
Preview endpoint does not trigger any notifications.

Actually, something I did not test is if a public link is loaded in which there is a file with no preview yet. Then opening the folder may trigger the preview computing and a download activity. But I’m not 100% sure that is possible in real life. How do you share a folder without triggering preview caching?

come-nc committed:
feat: add integration test for sharing activity

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
feat: Add context and test steps for activity in sharing

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
fix(sharebymail): Fix activity rich object id type

Fixes "Object for placeholder file is invalid, value 21 for key id is not a string"
Exception spotted in tests logs.

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
fix(tests): Sort activities by id to get the last one

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

nickvergessen commented on build/integration/features/bootstrap/BasicStructure.php:
```diff
@@ -192,8 +192,8 @@ protected function sendRequestForJSON(string $verb, string $url, $body = null, a
 		$options = [];
 		if ($this->currentUser === 'admin') {
 			$options['auth'] = ['admin', 'admin'];
-		} elseif (strpos($this->currentUser, 'guest') !== 0) {
-			$options['auth'] = [$this->currentUser, self::TEST_PASSWORD];
```
Can we keep this unchanged to not break apps that use integration test base from server?

come-nc replied:
No that breaks it. Can you explain how to fix it then? Why is sendingToWith using $this->regularUser but this one use another password?

nickvergessen replied:
I mixed it up I think. Seems everyone is only using the sendingToWith

nickvergessen replied:
Well now it says:
Fatal error: Undefined constant SharingContext::TEST_PASSWORD (Behat\Testwork\Call\Exception\FatalThrowableError)

So that's something different?
The actual fix is replacing `strpos() !== 0`. It does not account for `=== false` when it does not contain the word guest at all

come-nc replied:
No, `!==0` is on purpose. The idea is that if the user name starts with X, we do not set auth. But it’s anonymous in the other methods and guest in this one. But TEST_PASSWORD does not exist anyway.

I removed the last commit as it was breaking.

Do you know of applications using this method or is that theoretical?

nickvergessen replied:
Marked it as resolved already, not aware of the JSON method being used
